### ISSUE TITLE
some question about the learning rate

### ISSUE THREAD

**Issue by Cameltr** [MarcBotet/hamlet#9]
Hello, I would like to ask about the learning rate during the online training stage. The learning rate equation 6 mentioned in the paper is as follows. 
![image](https://github.com/MarcBotet/hamlet/assets/66940842/f3452eaf-c3d8-4f2e-8b8c-dea834b0955e)
During the first training domain (clear, 3 epoches), i.e. $B_i = B_{source}$ , the learning rate K is equal to $K_{min}$. According to the supplementary materials, $K_{min}$ is $1.5 \times 10^{-5}$.

But why the learning rate is 0 in the first three epochs in the experimental results graph reported in the Increasing Storm in the paper? 
![image](https://github.com/MarcBotet/hamlet/assets/66940842/6d29ea35-7956-482f-a5d9-0adfa57efe2f)

The following is my learning rate and mIoU result on the Incremental Storm dataset, which did not meet the results reported in the paper. I would appreciated if you could answer my question! 
![image](https://github.com/MarcBotet/hamlet/assets/66940842/18752366-92a5-46a7-bda9-ee7ee3757780)

**Comment by MarcBotet** [comment]
Hello, thanks for your interest in our work.

We designed Hamlet to start each scenario without training, only in eval mode. We only start the first training when we get a domain change. Since the source dataset is known, one can start a scenario with the needed information to detect a domain change. One easy way to implement this is to start each scenario with the "clear" scenario, this makes the first iterations show learning rate 0 since we are not training only evaluating (until it detects a domain change).

**Comment by Cameltr** [comment]
Thank you for your answer! I have successfully reproduced the corresponding mIoU results on the Increasing Storm dataset. Great job! By the way, I have some other questions below.

1. Is the GFlops of the model calculated using the code in `mmseg/tools/get_flops.py`?

2. is the `fps` calculated by using the `fps_array.npy` after training?

3. The fog dataset with different visible ranges (50m, 75m, 375m...) seems to only have a training set and no validation set. Did you manually partition it yourself?

**Comment by MarcBotet** [comment]
1- The GFlops from the paper are calculated manually after a full run. With the get_flops.py we obtained the flops of the full model, for each sub-module in the encoder, for the decoder, and for the light-weight decoder. Then, we check at each iteration step if it was an eval or training step and which sub-module has been used for training. With all this information we get the full training GFlops and average for total iterations. We follow the methodology in https://openai.com/blog/ai-and-compute/ to calculate the backpropagation FLOPs by multiplying the forward FLOPs by 2.
I am sorry but I cannot provide you with the script used to perform these calculations.

2- fps_array.npy contains the time taken for each step (not including data loading).  Since we use batch size 1, the fps can be obtained by getting the mean and 1/mean(fps_array)

3- Yes, the fog dataset was manually partitioned. For a fair comparison, we partitioned the fog dataset using the same partition as Onda. I do not have right now the exact partition. You can contact Onda authors to ask for it. I will update the issue when I have the partition list, but most probably it will take me a week.

**Comment by Cameltr** [comment]
Thanks a lot for your careful answer! Wish you all the best in your future work and life!

**Comment by Cameltr** [comment]
1. Hello, during testing in the foggy domains，may I ask if the source domain pretrained model `mitb1_uda.pth` was trained using **2475 partitioned cityscapes train set** instead of using the **original 2975 training dataset** ? I found using the provided `mitb1_uda.pth` during testing can cause overfitting.

2. As for the SHIFT dataset, did you manually partition it based on the provided Excel file `seq.csv` for the 6 scenarios?

**Comment by MarcBotet** [comment]
Hi,
1- the source domain was trained only on the partitioned train set, the other one is only kept for evaluation and is not included during training at all
2- Note that for shift we use the discrete scenario not the continuous one since it has too few images for our use case. We use the validation partition for validation

**Comment by Cameltr** [comment]
Thank you for your patient answer! I would like to ask if it is possible to provide a configuration file for model training on the source domain. I was unable to train a source domain model with good results based on the provided `config.py` and `experiment.py`. The main reason is that the configuration about **modular training**  resulted in incorrect weight of the lightweight decoder. I would greatly appreciate it if you could help me. It would be of great help to my graduation project!

**Comment by MarcBotet** [comment]
Sadly I don't have access anymore to the other configuration files. To train your model on the source domain you need to change the config file to "source-only" and change the appropriate parameters: a learning rate of 6×10−5, warm-up and linear decay scheduling, no modular training and batch size 4. The easier way is to check DAFormer config file for training the source, since we followed the same procedure (but increasing batch size to 4). One different thing is that we start from a trained SegFormer on cityscapes weights instead on imagenet weights

**Comment by Cameltr** [comment]
> Sadly I don't have access anymore to the other configuration files. To train your model on the source domain you need to change the config file to "source-only" and change the appropriate parameters: a learning rate of 6×10−5, warm-up and linear decay scheduling, no modular training and batch size 4. The easier way is to check DAFormer config file for training the source, since we followed the same procedure (but increasing batch size to 4). One different thing is that we start from a trained SegFormer on cityscapes weights instead on imagenet weights

Thank you for your nice answer, which means that Hamlet uses a source pre-trained model and then trains this lightweight decoder separately?

**Comment by MarcBotet** [comment]
yes, we use pre-trained on cityscapes (finetuned to work properly with our image resolution since is different from SegFormer original code) and then we freeze everything to train the lightweight decoder. You can check the code for freezing here /mmseg/models/decode_heads/segformer_head.py
